### ISSUE TITLE
Dune build: documentation in Makefile.dune seems to be outdated / misleading

### ISSUE THREAD

**Issue by MSoegtropIMC** [coq/coq#12167]
The documentation in Makefile.dune includes:
```
	@echo " To run an app \\in {coqc,coqtop,coqbyte,coqide}:"
	@echo ""
	@echo "  - use 'dune exec -- dev/shim/app-prelude args'"
	@echo ""
```
This seems to be outdated - the file dev/shim/app-prelude does not exist and the file dev/shim/dune seems to have a different purpose.

This opens the question what to do with files built with:
```
./configure -local
make -f Makefile.dune world
```
I couldn't find a way to run such files.

I also tried following
```
	@echo " To build a package, you can use:"
	@echo ""
	@echo "  - 'dune build package.install' : build package in developer mode"
	@echo "  - 'dune build -p package' : build package in release mode"
```
the command
```
dune build -p coq
```
This produces a coqc in `_build/install/default/bin/coqc` which exits with an error that it can't find the library. The command
```
dune build -p coqide
```
seems to delete the coqc built with the previous command (and everything else required to run coqide) so that doesn't seem to be useful either.

In summary, following the documentation in Makefile.dune doesn't lead to a runnable coqc or coqide.

I am using coq master as of today, `ac7996dbd4e6d49d7b5416b27d63b8ce9814f211`.

**Comment by SkySkimmer** [comment]
`app` is one of `coqc` `coqtop` etc

Don't use configure with dune

**Comment by Zimmi48** [comment]
The documentation is not outdated. Just try `dune exec -- dev/shim/coqtop-prelude`. That being said, this is a developer target and should not be used when the intent is to create a package.

`dune build -p coq` and `dune build -p coqide` are the ones to use when intending to install. Use `dune install` afterwards (cf. `dune install --help`).

**Comment by ejgallego** [comment]
Yes, maybe we should use `$app \in coqc`, you are not supposed to use app directly.

I have `dune coqtop` / `dune coqc` almost ready anyways, I have the impression this will help a lot of people.

> ```
> dune build -p coqide
> ```
> 
> seems to delete the coqc built with the previous command (and everything else required to run coqide) so that doesn't seem to be useful either.

We could maybe add some more documentation on `-p` ? When you use `-p` a package will be built,  but indeed in this case the local dependencies will be ignored, so in this case `coqide` will be built against the global `coq` package. For developers it is better to use the shims IMHO.

> Don't use configure with dune

Note that it is supported , and for example configure must be called to set `-prefix` properly in packages for now. But you are right, for a dev setup no need to call it.

**Comment by MSoegtropIMC** [comment]
Indeed I like dune because it builds fast, but each time I try it, I can't figure it out.

> Yes, maybe we should use `$app \in coqc`, you are not supposed to use "app" directly, it works as a variable here.

Without any mark that `app` is a variable in `dev/shim/app-prelude`, I did assume that `app` is part of the args. A common "syntax" would also be `dev/shim/<app>-prelude`. What also is confusing is that this looks like as if there should be shell scripts (or whatever) named dev/shim/<app>-prelude but there are not. So I think part of the documentation problem is the obscurity of the mechanism.

> We could maybe add some more documentation on -p ?

It would help to add documentation on how to build and install coqc and coqide and how to control where things are installed. The fact that a `dune build -p coqide` destroys the result of `dune build -p coq`. I would be tempted to first call both build commands and then both install commands.

**Comment by ejgallego** [comment]
> Without any mark that `app` is a variable in `dev/shim/app-prelude`, I did assume that `app` is part of the args. A common "syntax" would also be `dev/shim/<app>-prelude`. What also is confusing is that this looks like as if there should be shell scripts (or whatever) named dev/shim/-prelude but there are not. So I think part of the documentation problem is the obscurity of the mechanism.

Well, the same could be said for `dune exec coqc` , when you call that command `coqc` still doesn't exists.

> It would help to add documentation on how to build and install coqc and coqide and how to control where things are installed. The fact that a `dune build -p coqide` destroys the result of `dune build -p coq`. I would be tempted to first call both build commands and then both install commands.

On the other hand this is pretty well documented in Dune's own documentation, so I'm not sure we should just go and duplicate it vs users reading the real doc.

If you want to build two packages at the same time you use `dune build -p coq,coqide`

**Comment by MSoegtropIMC** [comment]
> On the other hand this is pretty well documented in Dune's own documentation, so I'm not sure we should just go and duplicate it vs users reading the real doc.

I think it is problematic if a build system can only be used when reading the docs. For just using make with predefined make files it is usually sufficient to just know the targets and a few very basic command line options. For dune this doesn't seem to be the case, which is of cause partly also because people are more used to make. But I think the popularity of dune will increase faster, if the command lines for the say 5..10 most common use cases are explicitly given here (developer build, local install, system install, with/without coqide).

I must also say that the dune documentation is not well suited for people who just want to use it with existing dune files.

**Comment by ejgallego** [comment]
> I think it is problematic if a build system can only be used when reading the docs.

Here we are talking about package generation; if you are a packager indeed I think you should read the docs about how to generate packages.



> For just using make with predefined make files it is usually sufficient to just know the targets and a few very basic command line options. For dune this doesn't seem to be the case

I'm afraid I cannot agree here, even if the documentation can be improved, I find the current provided documentation pretty clear and most use cases are already covered. So far the only problem here is that it wasn't clear that `app` was playing the role of a variable.

> explicitly given here (developer build, local install, system install, with/without coqide).

These are already provided.

**Comment by ejgallego** [comment]
Moreover all this is pretty well documented in `dev/doc/build-system.dune` and install.dune.

**Comment by MSoegtropIMC** [comment]
> Moreover all this is pretty well documented in `dev/doc/build-system.dune` and install.dune.

All I can say is that I typically get along quite well with most sorts of computer systems, but dune is an exception for a reason I can't tell. Let me hide my own notes and try again - maybe I can find out what it is.

**Comment by MSoegtropIMC** [comment]
@ejgallego: I had a look at the provided documentation. My summary:

- `dev/doc/build-system.dune` is not where I would look for it. Building is not per se a "developer" job, so I looked under doc, not under dev/doc. Since there are quite a few dune files, it is a bit hard to search for dune specific stuff, so I overlooked this file. I would recommend to put a reference to `dev/doc/build-system.dune` into `Makefile.dune` at a prominent place (maybe in one of the first lines of the echo block). Since `Makefile.dune` contains quite a bit of documentation users might think that this is it.

- The file `dev/doc/build-system.dune` is indeed quite informative. Just the section `## Compositionality, developer and release modes.` I don't understand (at all).

- The question on how to make a local and/or global install of coq and/or coqide is I think not sufficiently answered in the docs - maybe this has to do with the previous point (that is it is explained but I don't understand the explanation). I think a command line example for making a build + local install, build + system install, build + install into folder xyz would help.

- as discussed change the section in Makefile.dune:
```
	@echo " To run an app \\in {coqc,coqtop,coqbyte,coqide}:"
	@echo ""
	@echo "  - use 'dune exec -- dev/shim/app-prelude args'"
```
such that it is clear that the "variable" app is not part of args (as I assumed).

I think these minor changes should reduce the issues people might have with dune substantially.
